datetimeformat.html

```diff
@@ -81,6 +81,8 @@ <h1>BasicFormatMatcher ( _options_, _formats_ )</h1>
         1. Let _longMorePenalty_ be 6.
         1. Let _shortLessPenalty_ be 6.
         1. Let _shortMorePenalty_ be 3.
+        1. <ins>Let _offsetPenalty_ be 1.</ins>
+        1. <ins>Let _forbiddenPenalty_ be 1000.</ins>
```

Comment:
I think I would just re-use the existing offset values instead of inventing new ones.  In particular, I think 1000 is far too much.  I suppose you could make the argument for _offsetPenalty_ to be smaller than _shortMorePenalty_.

Reply:
1. yes, offsetPenalty need to be smaller than _shortMorePenalty_ otherwise the algorithm may use a different format than using a different timeZone fallback.
2. Basically, I need to prevent the algorithm to use another timeZone name since we should NOT fallback to them. If 1000 is far to much, then what it should be? It need to be bigger than _removalPenalty_ Maybe 240 ?

Reply:
> I need to prevent the algorithm to use another timeZone name since we should NOT fallback to them

Why is it so bad to fall back to a different width of time zone name?

If you must, I'd consider it no worse than an addition penalty plus removal penalty, which is 140 points.

Reply:
> Why is it so bad to fall back to a different width of time zone name?

Because ICU implementation won't....
see TimeZone::getDisplayName()
https://source.chromium.org/chromium/chromium/src/+/main:third_party/icu/source/i18n/timezone.cpp;l=1222?q=timezone%20Name%20third_party%2Ficu&ss=chromium

Reply:
cosign on the fallback to different width since that's more consistent with how other attributes are penalized.

Reply:
what do you mean? I don't think other attributes has the same kind of relationship between the fallback. Could you explain which attributes are similar to this kind of fallback condition? other attribute all either allow to fallback or the pattern do not have such element. But in this case, we are allow certain kind of fallback but not allow other kind of fallback. I don't think there are other attributes have the same kind of relationship we can consist with. Could you show me one I can follow?

Reply:
> Because ICU implementation won't....

This is in BasicFormatMatcher, which is not ICU. The ICU algorithm is implementation-defined in BestFormatMatcher.

Reply:
that is a good point. So... what is your suggestion of what BasicFormatMatcher should do? I try to write it in a way close to what ICU is doing. If we do not want to do so, what it should do instead? and why is that better?  

BTW, v8 currently only implement "best fit" matcher

Reply:
For BasicFormatMatcher, I think the simplest algorithm would be to essentially treat the three timeZoneName styles (default, generic, and offset) as separate fields; if you requested shortOffset but get longOffset, that should incur a length penalty, but if you requested shortOffset but get shortGeneric, that should incur a removed field penalty.

Reply:
How about if you request shortGeneric but get shortOffset?
Because UTS35 does mention the fallback from shortGeneric to shortOffset and from short to shortOffset

Reply:
If there's a UTS 35 algorithm, then I'd try to encode that algorithm here.  However, using costs like 1 and 1000 don't fit with the spirit of this algorithm, IMHO.  But, I don't hold a strong opinion, so I will unset the review bit.

Reply:
There are thing 1) mentioned in UTS35 and there are also 2) thing not mentioned in UTS35 but coded in ICU. My current PR is trying to fit 1) and 2). But since @sffc point out this BasicFormatMatcher not BestFormatMatcher I probably should change it in a way to only capture 1) and still fit reasonably in the frame work. so I like the idea that @sffc mention A) if you requested shortOffset but get longOffset, that should incur a length penalty, I also like B) requested shortOffset but get shortGeneric, that should incur a removed field penalty . In addition, I think I should make it so C) requested shortGeneric but get shortOffset a smaller fallback penalty should be less than length penalty . How about the following

Where _offsetPenalty_ is 1 and I reuse other penality.

| _optionsProp_  |  _formatProp_  |   |   |   |   |   |  
|---|---|---|---|---|---|---|
|                                | _short_  |  _shortGeneric_  |     _shortOffset_      | _long_  | _longGeneric_ |  _longOffset_ |
|  _short_                |            0  |  _removalPenalty_  |    _offsetPenalty_   |   _shortMorePenalty_   |  _removalPenalty_   |  _removalPenalty_      |  
|  _shortGeneric_  |  _removalPenalty_   |   0                            |    _offsetPenalty_   |  _removalPenalty_  |     _shortMorePenalty_     |  _removalPenalty_                        |  
|  _shortOffset_     | _removalPenalty_  |  _removalPenalty_   |                                0  |   _removalPenalty_  |  _removalPenalty_  |   _shortMorePenalty_      |
|  _long_                  |   _longLessPenalty_   | _removalPenalty_   | _removalPenalty_   |       0       | _removalPenalty_     |   _offsetPenalty_  |
|  _longGeneric_    |  _removalPenalty_  |    _longLessPenalty_  | _removalPenalty_  |  _removalPenalty_   |             0             |   _offsetPenalty_   |
|  _longOffset_       | _removalPenalty_   |  _removalPenalty_  | _longLessPenalty_    |   _removalPenalty_ |  _removalPenalty_    |             0              |

Reply:
That table looks good to me. Thanks!

One question: if you asked for *short* or *shortGeneric* but get *longOffset*, perhaps that should be penalized by _offsetPentalty_ + _shortMorePenalty_ rather than _removalPenalty_.  Likewise for *long* and *longGeneric* with *shortOffset*.

Reply:
could you please take another look of the table.

Reply:
Perfect, thanks!